src/Pirouette/Term/Symbolic/Eval.hs

```diff
@@ -273,25 +352,50 @@ symeval' t@(R.App hd args) = do
       --
       -- Naturally, the first path is already impossible so we do not need to
       -- move on to evaluate N.
-      DDestructor tyName -> do
-        Just (UnDestMeta _ _ tyParams term _ cases _) <- lift $ runMaybeT (unDest t)
-        Datatype _ _ _ consList <- lift $ prtTypeDefOf tyName
+      Just (DDestructor tyName) -> do
+        Just (UnDestMeta _ _ tyParams term tyRes cases excess) <- lift $ lift $ runMaybeT (unDest t)
+        Datatype _ _ _ consList <- lift $ lift $ prtTypeDefOf tyName
         -- We know what is the type of all the possible term results, its whatever
         -- type we're destructing applied to its arguments, making sure it contains
         -- no meta variables.
         let tyParams' = map typeFromMeta tyParams
-        term' <- symeval term
-        asum $
+        (term', somethingWasEvaluated) <- listen $ symEvalOneStep term
+        -- We only do the case distinction if we haven't taken any step
+        -- in the previous step. Otherwise it wouldn't be a "one step" evaluator.
+        if somethingWasEvaluated == Any True
```

Comment:
This is some funny business... :thinking: 

So, for builtins and destructors, we evaluate them in a call-by-value fashion; which might well make sense. I just wonder whether the performance will be penalized by evaluating inside an argument then reconstructing the whole term, just to have to have to traverse that term and do another step within it on the next iteration.

I don't think its worth changing it right now, but it is worth keeping it in our minds. I can see how `symEvalOneStep` could perhaps become `stmEvalSomeSteps` instead, and in those cases it just evaluates the argument a little more eagerly, perhaps until the first constructor is at the head, so pattern matching can go forward.

Reply:
Indeed, there might be some exploration to be done here. The three knobs I can think of are:

- evaluating the head only as needed for pattern matching;
- whether to use a call-by-value or call-by-name strategy in application;
- whether to go into recursive operations.